### PR TITLE
feat(kubernetes): Create keyword and network policy edge builders

### PR DESCRIPTION
**Pull request by PelegLi**
**By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

## Description

- Added `KeywordEdgeBuilder` class for building edges for resources with specific named properties defined in `ResourceKeywordIdentifier`
- Added `NetworkPolicyEdgeBuilder` class for building edges for NetworkPolicy resources (with Pods)
- Updated `_get_k8s_block_metadata` to support creating labels for `Service` resources
- Added `FILTERED_RESOURCES_FOR_EDGE_BUILDERS` to filter out resources with specific strategy from using builders with a generic strategy (e.g: filter `NetworkPolicy` from `LabelSelectorEdgeBuilder`)
- Added unit tests

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my feature, policy, or fix is effective and works
- [x] New and existing tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: checkov/kubernetes/graph_builder/graph_components/edge_builders/KeywordEdgeBuilder.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+                            reference_value_item = f".{i}.".join(reference_value_item.split(".", 1))
+                        vertex_ref = vertex.attributes.get(reference_value_item)
+                        potential_vertex_ref = potential_vertex.attributes.get(reference_key)
+                        if vertex_ref is None or potential_vertex_ref is None or vertex_ref != potential_vertex_ref:
+                            # if not all attributes match then it's not qualified as an edge
+                            match = False
+                            break
+                if match:
+                    connections.append(potential_vertex_index)
+                    break
```
Comment (rotemavni) [inline checkov/kubernetes/graph_builder/graph_components/edge_builders/KeywordEdgeBuilder.py:54]: If I understand the logic correctly, as long as we don't match all attributes, we want to append to the connection?

Comment (PelegLi) [reply checkov/kubernetes/graph_builder/graph_components/edge_builders/KeywordEdgeBuilder.py:54]: Only if we match all attributes for a single object in `KINDS_KEYWORDS_MAP[resource]` list - we append a connection

File: checkov/kubernetes/graph_builder/graph_components/edge_builders/KeywordEdgeBuilder.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+                for reference_key, reference_value in references_definition.items():
+                    reference_is_list = True if isinstance(reference_value, list) else False
+                    reference_values_list = force_list(reference_value)
+                    # some items are nested in lists and their value in the vertex is concatenated with their index,
+                    # like so:  subjects.0.name
+                    # the following lines force all properties to a list for ease of use and concatenate the index
+                    # only to items that are actual lists
+                    for i, reference_value_item in enumerate(reference_values_list):
+                        if reference_is_list:
+                            reference_value_item = f".{i}.".join(reference_value_item.split(".", 1))
```
Comment (YaaraVerner) [inline checkov/kubernetes/graph_builder/graph_components/edge_builders/KeywordEdgeBuilder.py:45]: As we talked about, you should run on the list of values from the vertex itself

File: checkov/kubernetes/graph_builder/graph_components/edge_builders/LabelSelectorEdgeBuilder.py [hunk 3]
```diff
@@ -1,16 +1,17 @@
...
 
 class LabelSelectorEdgeBuilder(K8SEdgeBuilder):
 
     @staticmethod
     def should_search_for_edges(vertex: KubernetesBlock) -> bool:
-        if vertex.metadata is not None and vertex.metadata.labels is not None:
-            return True
-        return False
+        return vertex.metadata is not None \
+            and vertex.metadata.labels is not None \
```
Comment (YaaraVerner) [inline checkov/kubernetes/graph_builder/graph_components/edge_builders/LabelSelectorEdgeBuilder.py:13]: will the metadata always include labels?

**Review by YaaraVerner — COMMENTED** [review]
Nice work! 👑

File: checkov/kubernetes/graph_builder/graph_components/ResourceKeywordIdentifier.py [hunk 4]
```diff
@@ -1,9 +1,30 @@
...
     KINDS_KEYWORDS_MAP = {
-        "PersistentVolumeClaim": ["claimName"],
-        "ServiceAccount": ["serviceAccountName"],
-        "ClusterRole": ["rules[].resources", "rules[].resourceNames"],
-        "ClusterRoleBinding": [["roleRef.name", "kind"], ["subjects[].name", "kind"], ["aggregationRule.clusterRoleSelectors.matchLabels"]],
-        "Role": ["rules[].resources", "rules[].resourceNames"],
-        "RoleBinding": [["roleRef.name", "kind"], ["subjects[].name", "kind"]]
+        # "PersistentVolumeClaim": ["claimName"],
+        # "ClusterRole": ["rules[].resources", "rules[].resourceNames"],
+        # "Role": ["rules[].resources", "rules[].resourceNames"],
```
Comment (gruebel) [inline checkov/kubernetes/graph_builder/graph_components/ResourceKeywordIdentifier.py:18]: can this be removed or is it left for tackling later?

File: checkov/kubernetes/graph_builder/graph_components/edge_builders/LabelSelectorEdgeBuilder.py [hunk 5]
```diff
@@ -1,16 +1,17 @@
...
+            and vertex.attributes.get("kind") not in FILTERED_RESOURCES_FOR_EDGE_BUILDERS
```
Comment (gruebel) [inline checkov/kubernetes/graph_builder/graph_components/edge_builders/LabelSelectorEdgeBuilder.py:14]: does it make sense to search for edges if we don't have `kind`? if we assume it is always there, then directly access it `vertex.attributes["kind"]`

File: checkov/kubernetes/graph_builder/local_graph.py [hunk 6]
```diff
@@ -104,29 +104,29 @@ def _create_edge(self, origin_vertex_index: int, dest_vertex_index: int, label:
...
         spec = resource.get('spec')
+        match_labels: Dict[str, Any] | None = None
         if isinstance(spec, list):
             for spec_item in spec:
                 if spec_item.get('selector'):
-                    match_labels = spec_item.get('selector').get('matchLabels')
+                    if resource.get('kind') == "Service":
+                        match_labels = spec_item.get('selector')
+                    else:
+                        match_labels = spec_item.get('selector').get('matchLabels')
```
Comment (gruebel) [inline checkov/kubernetes/graph_builder/local_graph.py:114]: maybe extract it to a small function, because you are using it two times?

**Review by gruebel — COMMENTED** [review]
great work 💪

File: checkov/kubernetes/graph_builder/graph_components/ResourceKeywordIdentifier.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (PelegLi) [reply checkov/kubernetes/graph_builder/graph_components/ResourceKeywordIdentifier.py:18]: It's a TODO place holder for resources that will probably need to be handled later on

File: checkov/kubernetes/graph_builder/graph_components/edge_builders/LabelSelectorEdgeBuilder.py (hunk 3 above)
Comment (PelegLi) [reply checkov/kubernetes/graph_builder/graph_components/edge_builders/LabelSelectorEdgeBuilder.py:13]: yes, it's initialized with the Metadata class

File: checkov/kubernetes/graph_builder/graph_components/ResourceKeywordIdentifier.py (hunk 4 above)
Comment (PelegLi) [reply checkov/kubernetes/graph_builder/graph_components/ResourceKeywordIdentifier.py:18]: yes, it's a TODO place holder for resources that will probably need to be handled later on

File: checkov/kubernetes/graph_builder/graph_components/edge_builders/LabelSelectorEdgeBuilder.py (hunk 5 above)
Comment (PelegLi) [reply checkov/kubernetes/graph_builder/graph_components/edge_builders/LabelSelectorEdgeBuilder.py:14]: unfortunately not all object have this attribute, hence [this](https://github.com/bridgecrewio/checkov/pull/3645)

File: checkov/kubernetes/graph_builder/graph_components/ResourceKeywordIdentifier.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (YaaraVerner) [reply checkov/kubernetes/graph_builder/graph_components/ResourceKeywordIdentifier.py:18]: so can you create indicated TODO comment for it?

File: checkov/kubernetes/graph_builder/graph_components/edge_builders/LabelSelectorEdgeBuilder.py (hunk 5 above)
Comment (gruebel) [reply checkov/kubernetes/graph_builder/graph_components/edge_builders/LabelSelectorEdgeBuilder.py:14]: ok, then wouldn't it be better to do this
```suggestion
            and "kind" in vertex.attributes
            and vertex.attributes["kind"] not in FILTERED_RESOURCES_FOR_EDGE_BUILDERS
```
just want to make sure, if we need to consider nodes, which have no `kind` attribute

**Review by gruebel — APPROVED** [review]
☕

**Pull request merged**